File: python/src/afd/core/handoff.py
"""Handoff types for AFD commands.

The HandoffResult type is returned by commands that bootstrap streaming
connections, allowing clients to connect to specialized protocols
(WebSocket, WebRTC, SSE, etc.) for real-time communication.

Example:
    >>> from afd.core.handoff import HandoffResult, is_handoff
    >>>
    >>> # Check if a result is a handoff
    >>> result = {"protocol": "websocket", "endpoint": "wss://example.com/chat"}
    >>> if is_handoff(result):
    ...     print(f"Connect to {result['endpoint']}")
"""

from typing import Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, Field


# Type alias for standard protocols
HandoffProtocol = Union[
    Literal["websocket", "webrtc", "sse", "http-stream"],
    str,  # Allow custom protocols
]


class ReconnectPolicy(BaseModel):
    """Reconnection policy for handoff connections.

    Attributes:
        allowed: Whether reconnection is allowed.
        max_attempts: Maximum number of reconnection attempts.
        backoff_ms: Base backoff time in milliseconds.

    Example:
        >>> policy = ReconnectPolicy(allowed=True, max_attempts=5, backoff_ms=1000)
    """

    allowed: bool
    max_attempts: Optional[int] = Field(default=None, ge=0)
    backoff_ms: Optional[int] = Field(default=None, ge=0)


class HandoffCredentials(BaseModel):
    """Authentication credentials for the handoff connection.

    Attributes:
        token: Bearer token for authentication.
        headers: Additional headers to include in the connection.
        session_id: Session ID for correlation.

    Example:
        >>> creds = HandoffCredentials(
        ...     token="eyJhbGciOiJIUzI1NiIs...",
        ...     session_id="session-abc123",
        ...     headers={"X-Custom-Header": "value"},
        ... )
    """

    token: Optional[str] = None
    headers: Optional[Dict[str, str]] = None
    session_id: Optional[str] = None


class HandoffMetadata(BaseModel):
    """Metadata for client decision-making about the handoff.

    Attributes:
        expected_latency: Expected latency in ms (hint for client).
        capabilities: Capabilities the channel supports.
        expires_at: When the handoff credentials expire (ISO 8601).
        reconnect: Reconnection policy.
        description: Human-readable description of the handoff.

    Example:
        >>> metadata = HandoffMetadata(
        ...     expected_latency=50,
        ...     capabilities=["text", "typing-indicator", "presence"],
        ...     expires_at="2025-01-15T12:00:00Z",
        ...     reconnect=ReconnectPolicy(allowed=True, max_attempts=5, backoff_ms=1000),
        ...     description="Real-time chat connection",
        ... )
    """

    expected_latency: Optional[int] = Field(default=None, ge=0)
    capabilities: Optional[List[str]] = None
    expires_at: Optional[str] = None
    reconnect: Optional[ReconnectPolicy] = None
    description: Optional[str] = None


class HandoffResult(BaseModel):
    """Result returned by commands that hand off to specialized protocols.

    This type is used as the data payload in CommandResult[HandoffResult].
    Commands returning handoffs should be marked with `handoff=True` and
    tagged appropriately (e.g., 'handoff', 'handoff:websocket').

    Attributes:
        protocol: Protocol type for client dispatch.
        endpoint: Full URL to connect to.
        credentials: Authentication credentials for the handoff.
        metadata: Metadata for client decision-making.

    Example:
        >>> from afd import success
        >>>
        >>> # In a command handler
        >>> handoff = HandoffResult(
        ...     protocol="websocket",
        ...     endpoint="wss://chat.example.com/rooms/123",
        ...     credentials=HandoffCredentials(token="abc123"),
        ...     metadata=HandoffMetadata(
        ...         capabilities=["text", "presence"],
        ...         reconnect=ReconnectPolicy(allowed=True),
        ...     ),
        ... )
        >>> result = success(handoff.model_dump())
    """

    protocol: str = Field(..., min_length=1)
    endpoint: str = Field(..., min_length=1)
    credentials: Optional[HandoffCredentials] = None
    metadata: Optional[HandoffMetadata] = None


# ═══════════════════════════════════════════════════════════════════════════════
# TYPE GUARDS
# ═══════════════════════════════════════════════════════════════════════════════


def is_handoff(value: Any) -> bool:
    """Type guard to check if a value is a HandoffResult.

    Args:
        value: Value to check.

    Returns:
        True if value is a valid HandoffResult.

    Example:
        >>> result = {"protocol": "websocket", "endpoint": "wss://example.com/chat"}
        >>> is_handoff(result)
        True
        >>> is_handoff({"invalid": "data"})
        False
    """
    if not isinstance(value, dict):
        return False

    # Required fields: protocol and endpoint
    protocol = value.get("protocol")
    if not isinstance(protocol, str) or not protocol:
        return False

    endpoint = value.get("endpoint")
    if not isinstance(endpoint, str) or not endpoint:
        return False

    # Optional credentials validation
    credentials = value.get("credentials")
    if credentials is not None:
        if not isinstance(credentials, dict):
            return False

        token = credentials.get("token")
        if token is not None and not isinstance(token, str):
            return False

        session_id = credentials.get("session_id")
        if session_id is not None and not isinstance(session_id, str):
            return False

        headers = credentials.get("headers")
        if headers is not None and not isinstance(headers, dict):
            return False

    # Optional metadata validation
    metadata = value.get("metadata")
    if metadata is not None:
        if not isinstance(metadata, dict):
            return False

        expected_latency = metadata.get("expected_latency")
        if expected_latency is not None and not isinstance(expected_latency, (int, float)):
            return False

        capabilities = metadata.get("capabilities")
        if capabilities is not None and not isinstance(capabilities, list):
            return False

        expires_at = metadata.get("expires_at")
        if expires_at is not None and not isinstance(expires_at, str):
            return False

        description = metadata.get("description")
        if description is not None and not isinstance(description, str):
            return False

        reconnect = metadata.get("reconnect")
        if reconnect is not None:
            if not isinstance(reconnect, dict):
                return False

            allowed = reconnect.get("allowed")
            if not isinstance(allowed, bool):
                return False

            max_attempts = reconnect.get("max_attempts")
            if max_attempts is not None and not isinstance(max_attempts, (int, float)):
                return False

            backoff_ms = reconnect.get("backoff_ms")
            if backoff_ms is not None and not isinstance(backoff_ms, (int, float)):
                return False

    return True


def is_handoff_protocol(handoff: Dict[str, Any], protocol: str) -> bool:
    """Type guard to check if a HandoffResult uses a specific protocol.

    Args:
        handoff: HandoffResult dict to check.
        protocol: Protocol to check for.

    Returns:
        True if the handoff uses the specified protocol.

    Example:
        >>> handoff = {"protocol": "websocket", "endpoint": "wss://example.com"}
        >>> is_handoff_protocol(handoff, "websocket")
        True
        >>> is_handoff_protocol(handoff, "sse")
        False
    """
    return handoff.get("protocol") == protocol


# ═══════════════════════════════════════════════════════════════════════════════
# COMMAND TYPE GUARDS
# ═══════════════════════════════════════════════════════════════════════════════


def is_handoff_command(command: Any) -> bool:
    """Check if a command definition is a handoff command.

    A command is a handoff command if:
    - It has `handoff=True` property, OR
    - It has a 'handoff' tag

    Args:
        command: Command definition (dict or object with attributes) to check.

    Returns:
        True if the command is a handoff command.

    Example:
        >>> is_handoff_command({"handoff": True})
        True
        >>> is_handoff_command({"tags": ["handoff"]})
        True
        >>> is_handoff_command({})
        False
    """
    # Handle both dict and object with attributes
    if isinstance(command, dict):
        # Check explicit handoff property
        if command.get("handoff") is True:
            return True

        # Check for handoff tag
        tags = command.get("tags", [])
        if tags and "handoff" in tags:
            return True
    else:
        # Handle object with attributes (e.g., CommandDefinition)
        if getattr(command, "handoff", None) is True:
            return True

        tags = getattr(command, "tags", None) or []
        if "handoff" in tags:
            return True

    return False


def get_handoff_protocol(command: Any) -> Optional[str]:
    """Get the handoff protocol from a command definition.

    Returns the protocol in this priority order:
    1. Explicit `handoff_protocol` property
    2. Protocol from 'handoff:{protocol}' tag
    3. None if not a handoff command or no protocol specified

    Args:
        command: Command definition (dict or object with attributes) to check.

    Returns:
        The handoff protocol or None.

    Example:
        >>> get_handoff_protocol({"handoff": True, "handoff_protocol": "websocket"})
        'websocket'
        >>> get_handoff_protocol({"tags": ["handoff", "handoff:sse"]})
        'sse'
        >>> get_handoff_protocol({})
        None
    """
    # Not a handoff command
    if not is_handoff_command(command):
        return None

    # Handle both dict and object with attributes
    if isinstance(command, dict):
        # Check explicit handoff_protocol property first
        handoff_protocol = command.get("handoff_protocol")
        if handoff_protocol:
            return handoff_protocol

        # Check for handoff:{protocol} tag
        tags = command.get("tags", [])
    else:
        # Handle object with attributes
        handoff_protocol = getattr(command, "handoff_protocol", None)
        if handoff_protocol:
            return handoff_protocol

        tags = getattr(command, "tags", None) or []

    # Find handoff:{protocol} tag
    for tag in tags:
        if isinstance(tag, str) and tag.startswith("handoff:"):
            return tag[len("handoff:"):]

    return None


def create_handoff(
    protocol: str,
    endpoint: str,
    *,
    token: Optional[str] = None,
    session_id: Optional[str] = None,
    headers: Optional[Dict[str, str]] = None,
    expected_latency: Optional[int] = None,
    capabilities: Optional[List[str]] = None,
    expires_at: Optional[str] = None,
    reconnect_allowed: Optional[bool] = None,
    reconnect_max_attempts: Optional[int] = None,
    reconnect_backoff_ms: Optional[int] = None,
    description: Optional[str] = None,
) -> Dict[str, Any]:
    """Create a HandoffResult dict with the given parameters.

    This is a convenience function for creating handoff results in command handlers.

    Args:
        protocol: Protocol type (e.g., "websocket", "sse", "webrtc").
        endpoint: Full URL to connect to.
        token: Bearer token for authentication.
        session_id: Session ID for correlation.
        headers: Additional headers to include.
        expected_latency: Expected latency in ms.
        capabilities: Channel capabilities.
        expires_at: When credentials expire (ISO 8601).
        reconnect_allowed: Whether reconnection is allowed.
        reconnect_max_attempts: Max reconnection attempts.
        reconnect_backoff_ms: Base backoff time in ms.
        description: Human-readable description.

    Returns:
        A dict matching HandoffResult structure.

    Example:
        >>> from afd import success
        >>> handoff = create_handoff(
        ...     protocol="websocket",
        ...     endpoint="wss://chat.example.com/room/123",
        ...     token="abc123",
        ...     capabilities=["text", "presence"],
        ...     reconnect_allowed=True,
        ... )
        >>> result = success(handoff)
    """
    result: Dict[str, Any] = {
        "protocol": protocol,
        "endpoint": endpoint,
    }

    # Build credentials if any credential fields provided
    if token is not None or session_id is not None or headers is not None:
        credentials: Dict[str, Any] = {}
        if token is not None:
            credentials["token"] = token
        if session_id is not None:
            credentials["session_id"] = session_id
        if headers is not None:
            credentials["headers"] = headers
        result["credentials"] = credentials

    # Build metadata if any metadata fields provided
    has_metadata = any([
        expected_latency is not None,
        capabilities is not None,
        expires_at is not None,
        reconnect_allowed is not None,
        description is not None,
    ])

    if has_metadata:
        metadata: Dict[str, Any] = {}
        if expected_latency is not None:
            metadata["expected_latency"] = expected_latency
        if capabilities is not None:
            metadata["capabilities"] = capabilities
        if expires_at is not None:
            metadata["expires_at"] = expires_at
        if description is not None:
            metadata["description"] = description

        # Build reconnect policy if specified
        if reconnect_allowed is not None:
            reconnect: Dict[str, Any] = {"allowed": reconnect_allowed}
            if reconnect_max_attempts is not None:
                reconnect["max_attempts"] = reconnect_max_attempts
            if reconnect_backoff_ms is not None:
                reconnect["backoff_ms"] = reconnect_backoff_ms
            metadata["reconnect"] = reconnect

        result["metadata"] = metadata

    return result

File: packages/core/src/handoff.ts
/**
 * @fileoverview Handoff types for AFD commands
 *
 * The HandoffResult interface is returned by commands that bootstrap
 * streaming connections, allowing clients to connect to specialized
 * protocols (WebSocket, WebRTC, SSE, etc.) for real-time communication.
 */

// ═══════════════════════════════════════════════════════════════════════════════
// HANDOFF TYPES
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Protocol type for client dispatch.
 *
 * Standard protocols are provided, but custom protocols are allowed
 * via the string union type.
 */
export type HandoffProtocol = 'websocket' | 'webrtc' | 'sse' | 'http-stream' | string;

/**
 * Authentication credentials for the handoff connection.
 *
 * @example
 * ```typescript
 * const credentials: HandoffCredentials = {
 *   token: 'eyJhbGciOiJIUzI1NiIs...',
 *   sessionId: 'session-abc123',
 *   headers: { 'X-Custom-Header': 'value' }
 * };
 * ```
 */
export interface HandoffCredentials {
	/** Bearer token for authentication */
	token?: string;

	/** Additional headers to include in the connection */
	headers?: Record<string, string>;

	/** Session ID for correlation */
	sessionId?: string;
}

/**
 * Metadata for client decision-making about the handoff.
 *
 * @example
 * ```typescript
 * const metadata: HandoffMetadata = {
 *   expectedLatency: 50,
 *   capabilities: ['text', 'typing-indicator', 'presence'],
 *   expiresAt: '2025-01-15T12:00:00Z',
 *   reconnect: {
 *     allowed: true,
 *     maxAttempts: 5,
 *     backoffMs: 1000
 *   },
 *   description: 'Real-time chat connection'
 * };
 * ```
 */
export interface HandoffMetadata {
	/** Expected latency in ms (hint for client) */
	expectedLatency?: number;

	/** Capabilities the channel supports */
	capabilities?: string[];

	/** When the handoff credentials expire (ISO 8601) */
	expiresAt?: string;

	/** Reconnection policy */
	reconnect?: {
		/** Whether reconnection is allowed */
		allowed: boolean;
		/** Maximum number of reconnection attempts */
		maxAttempts?: number;
		/** Base backoff time in milliseconds */
		backoffMs?: number;
	};

	/** Human-readable description of the handoff */
	description?: string;
}

/**
 * Result returned by commands that hand off to specialized protocols.
 *
 * This type is used as the data payload in CommandResult<HandoffResult>.
 * Commands returning handoffs should be marked with `handoff: true` and
 * tagged appropriately (e.g., 'handoff', 'handoff:websocket').
 *
 * @example
 * ```typescript
 * const result = await client.call<HandoffResult>('chat.connect', { roomId });
 * if (result.success && result.data) {
 *   const ws = new WebSocket(result.data.endpoint);
 *   ws.send(JSON.stringify({ auth: result.data.credentials?.token }));
 * }
 * ```
 */
export interface HandoffResult {
	/** Protocol type for client dispatch */
	protocol: HandoffProtocol;

	/** Full URL to connect to */
	endpoint: string;

	/** Authentication credentials for the handoff */
	credentials?: HandoffCredentials;

	/** Metadata for client decision-making */
	metadata?: HandoffMetadata;
}

// ═══════════════════════════════════════════════════════════════════════════════
// TYPE GUARDS
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Type guard to check if a value is a HandoffResult.
 *
 * @param value - Value to check
 * @returns True if value is a HandoffResult
 *
 * @example
 * ```typescript
 * const result = await client.call('some.command', input);
 * if (result.success && isHandoff(result.data)) {
 *   // TypeScript knows result.data is HandoffResult
 *   connect(result.data.protocol, result.data.endpoint);
 * }
 * ```
 */
export function isHandoff(value: unknown): value is HandoffResult {
	if (typeof value !== 'object' || value === null) {
		return false;
	}

	const obj = value as Record<string, unknown>;

	// Required fields: protocol and endpoint
	if (typeof obj.protocol !== 'string' || obj.protocol.length === 0) {
		return false;
	}

	if (typeof obj.endpoint !== 'string' || obj.endpoint.length === 0) {
		return false;
	}

	// Optional credentials validation
	if (obj.credentials !== undefined) {
		if (typeof obj.credentials !== 'object' || obj.credentials === null) {
			return false;
		}
		const creds = obj.credentials as Record<string, unknown>;
		if (creds.token !== undefined && typeof creds.token !== 'string') {
			return false;
		}
		if (creds.sessionId !== undefined && typeof creds.sessionId !== 'string') {
			return false;
		}
		if (creds.headers !== undefined) {
			if (typeof creds.headers !== 'object' || creds.headers === null) {
				return false;
			}
		}
	}

	// Optional metadata validation
	if (obj.metadata !== undefined) {
		if (typeof obj.metadata !== 'object' || obj.metadata === null) {
			return false;
		}
		const meta = obj.metadata as Record<string, unknown>;
		if (meta.expectedLatency !== undefined && typeof meta.expectedLatency !== 'number') {
			return false;
		}
		if (meta.capabilities !== undefined && !Array.isArray(meta.capabilities)) {
			return false;
		}
		if (meta.expiresAt !== undefined && typeof meta.expiresAt !== 'string') {
			return false;
		}
		if (meta.description !== undefined && typeof meta.description !== 'string') {
			return false;
		}
		if (meta.reconnect !== undefined) {
			if (typeof meta.reconnect !== 'object' || meta.reconnect === null) {
				return false;
			}
			const reconnect = meta.reconnect as Record<string, unknown>;
			if (typeof reconnect.allowed !== 'boolean') {
				return false;
			}
			if (reconnect.maxAttempts !== undefined && typeof reconnect.maxAttempts !== 'number') {
				return false;
			}
			if (reconnect.backoffMs !== undefined && typeof reconnect.backoffMs !== 'number') {
				return false;
			}
		}
	}

	return true;
}

/**
 * Type guard to check if a HandoffResult uses a specific protocol.
 *
 * @param handoff - HandoffResult to check
 * @param protocol - Protocol to check for
 * @returns True if the handoff uses the specified protocol
 *
 * @example
 * ```typescript
 * if (isHandoffProtocol(handoff, 'websocket')) {
 *   const ws = new WebSocket(handoff.endpoint);
 * } else if (isHandoffProtocol(handoff, 'sse')) {
 *   const eventSource = new EventSource(handoff.endpoint);
 * }
 * ```
 */
export function isHandoffProtocol(handoff: HandoffResult, protocol: HandoffProtocol): boolean {
	return handoff.protocol === protocol;
}

// ═══════════════════════════════════════════════════════════════════════════════
// COMMAND TYPE GUARDS
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Minimal interface for checking if a command is a handoff command.
 * This avoids circular dependencies with commands.ts.
 */
interface HandoffCommandLike {
	handoff?: boolean;
	handoffProtocol?: string;
	tags?: string[];
}

/**
 * Check if a command definition is a handoff command.
 *
 * A command is a handoff command if:
 * - It has `handoff: true` property, OR
 * - It has a 'handoff' tag
 *
 * @param command - Command definition to check
 * @returns True if the command is a handoff command
 *
 * @example
 * ```typescript
 * const commands = registry.list();
 * const handoffCommands = commands.filter(isHandoffCommand);
 * ```
 */
export function isHandoffCommand(command: HandoffCommandLike): boolean {
	// Check explicit handoff property
	if (command.handoff === true) {
		return true;
	}

	// Check for handoff tag
	if (command.tags?.includes('handoff')) {
		return true;
	}

	return false;
}

/**
 * Get the handoff protocol from a command definition.
 *
 * Returns the protocol in this priority order:
 * 1. Explicit `handoffProtocol` property
 * 2. Protocol from 'handoff:{protocol}' tag
 * 3. undefined if not a handoff command or no protocol specified
 *
 * @param command - Command definition to check
 * @returns The handoff protocol or undefined
 *
 * @example
 * ```typescript
 * const protocol = getHandoffProtocol(chatConnectCommand);
 * if (protocol === 'websocket') {
 *   // Handle WebSocket handoff
 * }
 * ```
 */
export function getHandoffProtocol(command: HandoffCommandLike): HandoffProtocol | undefined {
	// Not a handoff command
	if (!isHandoffCommand(command)) {
		return undefined;
	}

	// Check explicit handoffProtocol property first
	if (command.handoffProtocol) {
		return command.handoffProtocol;
	}

	// Check for handoff:{protocol} tag
	const protocolTag = command.tags?.find((tag) => tag.startsWith('handoff:'));
	if (protocolTag) {
		return protocolTag.slice('handoff:'.length);
	}

	return undefined;
}
